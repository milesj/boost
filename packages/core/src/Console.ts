/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import exit from 'exit';
import cliSize from 'term-size';
import ansiEscapes from 'ansi-escapes';
import Emitter from './Emitter';
import Tool from './Tool';
import { Debugger } from './types';
import Output from './Output';

// 16 FPS (60 FPS is actually too fast as it tears)
export const FPS_RATE = 62.5;

// Bind our writable streams for easy access
export const BOUND_WRITERS = {
  stderr: process.stderr.write.bind(process.stderr),
  stdout: process.stdout.write.bind(process.stdout),
};

export const WRAPPED_STREAMS = {
  stderr: false,
  stdout: false,
};

export type StreamType = 'stderr' | 'stdout';

export interface ConsoleState {
  disabled: boolean;
  final: boolean;
  started: boolean;
  stopped: boolean;
}

export default class Console extends Emitter {
  bufferedStreams: (() => void)[] = [];

  debug: Debugger;

  errorLogs: string[] = [];

  logs: string[] = [];

  outputQueue: Output[] = [];

  tool: Tool<any>;

  protected renderTimer: NodeJS.Timer | null = null;

  protected restoreCursorOnExit: boolean = false;

  protected state: ConsoleState = {
    disabled: false,
    final: false,
    started: false,
    stopped: false,
  };

  private writers: typeof BOUND_WRITERS;

  constructor(tool: Tool<any>, /* test only */ testWriters: typeof BOUND_WRITERS = BOUND_WRITERS) {
    super();

    this.debug = tool.createDebugger('console');
    this.tool = tool;
    this.writers = testWriters;

    // istanbul ignore next
    if (process.env.NODE_ENV !== 'test') {
      process
        .on('SIGINT', this.handleSignal)
        .on('SIGTERM', this.handleSignal)
        .on('uncaughtException', this.handleFailure)
        .on('unhandledRejection', this.handleFailure);
    }
  }

  /**
   * Disable the render loop entirely.
   */
  disable(): this {
    this.state.disabled = true;

    return this;
  }

  /**
   * Display a footer after all output.
   */
  displayFooter() {
    const { footer } = this.tool.options;

    if (footer) {
      this.out(footer, 1);
    }
  }

  /**
   * Display a header before all output.
   */
  displayHeader() {
    const { header } = this.tool.options;

    if (header) {
      this.out(header, 1);
    }
  }

  /**
   * Enable the render loop.
   */
  enable(): this {
    this.state.disabled = false;

    return this;
  }

  /**
   * Write a message to `stderr` with optional trailing newline(s).
   */
  err(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stderr(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Flush buffered stream output.
   */
  flushBufferedStreams(): this {
    this.bufferedStreams.forEach(buffer => {
      buffer();
    });

    return this;
  }

  /**
   * Flush the top output block in the queue.
   */
  flushOutputQueue(): this {
    const outputs = this.outputQueue.filter((out, i) => i === 0 || out.isConcurrent());

    // Erase the previous output
    outputs.forEach(output => {
      output.erasePrevious();
    });

    // Write buffered output
    this.flushBufferedStreams();

    // Write the next output
    outputs.forEach(output => {
      output.render();
    });

    // Remove completed outputs
    this.outputQueue = this.outputQueue.filter(out => !out.isComplete());

    // Stop the render loop once the queue is empty
    if (this.isEmptyQueue()) {
      this.stopRenderLoop();
    } else {
      this.startRenderLoop();
    }

    return this;
  }

  /**
   * Handle uncaught exceptions and unhandled rejections that bubble up.
   */
  handleFailure = (error: Error) => {
    this.start();
    this.debug('Uncaught exception or unresolved promise handled');
    this.stop(error);

    exit(2);
  };

  /**
   * Handle SIGINT and SIGTERM interruptions.
   */
  handleSignal = () => {
    this.start();
    this.debug('SIGINT or SIGTERM handled');
    this.stop(new Error(this.tool.msg('errors:processTerminated')));

    exit(2);
  };

  /**
   * Hide the console cursor.
   */
  hideCursor(): this {
    this.out(ansiEscapes.cursorHide);

    if (!this.restoreCursorOnExit && !this.isSilent()) {
      this.restoreCursorOnExit = true;

      // istanbul ignore next
      process.on('exit', () => {
        process.stdout.write(ansiEscapes.cursorShow);
      });
    }

    return this;
  }

  /**
   * Return true if the render loop has been disabled.
   */
  isDisabled(): boolean {
    return this.state.disabled;
  }

  /**
   * Return true if the output queue is empty.
   */
  isEmptyQueue(): boolean {
    return this.outputQueue.length === 0;
  }

  /**
   * Return true if the final render.
   */
  isFinalRender(): boolean {
    return this.state.final;
  }

  /**
   * Return true if the there should be no output.
   */
  isSilent(): boolean {
    return this.tool.config.silent;
  }

  /**
   * Return true if the defined stream has been wrapped by the console layer.
   */
  isStreamWrapped(type: StreamType): boolean {
    return !!WRAPPED_STREAMS[type];
  }

  /**
   * Log a message to display on success during the final render.
   */
  log(message: string): this {
    this.logs.push(message);

    return this;
  }

  /**
   * Log a live message to display during the rendering process.
   */
  logLive(message: string): this {
    // Write to the wrapped buffer
    process.stdout.write(message);

    return this;
  }

  /**
   * Log an error message to display on failure during the final render.
   */
  logError(message: string): this {
    this.errorLogs.push(message);

    return this;
  }

  /**
   * Write a message to `stdout` with optional trailing newline(s).
   */
  out(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stdout(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Enqueue a block of output to be rendered.
   */
  render(output: Output): this {
    if (this.isDisabled()) {
      throw new Error(
        'Output cannot be enqueued as the render loop has been disabled. This is usually caused by conflicting reporters.',
      );
    }

    if (!this.outputQueue.includes(output)) {
      this.outputQueue.push(output);
    }

    // Only run the render loop when output is enqueued
    this.startRenderLoop();

    return this;
  }

  /**
   * Handle the final rendering of all output before stopping.
   */
  renderFinalOutput(error: Error | null) {
    this.debug('Rendering final console output');
    this.state.final = true;

    // Mark all output as final
    this.outputQueue.forEach(output => {
      output.enqueue(true);
    });

    // Recursively render the remaining output
    while (!this.isEmptyQueue()) {
      this.flushOutputQueue();
    }

    // Stop the render loop
    this.stopRenderLoop();

    if (error) {
      if (this.errorLogs.length > 0) {
        this.err(`\n${this.errorLogs.join('\n')}\n`);
      }

      this.emit('error', [error]);
    } else {
      if (this.logs.length > 0) {
        this.out(`\n${this.logs.join('\n')}\n`);
      }

      this.displayFooter();
    }

    // Flush any stream output that still exists
    this.flushBufferedStreams();

    // Show the cursor incase it has been hidden
    this.out(ansiEscapes.cursorShow);
  }

  /**
   * Reset the cursor back to the bottom of the console.
   */
  resetCursor(): this {
    this.out(ansiEscapes.cursorTo(0, cliSize().rows));

    return this;
  }

  /**
   * Show the console cursor.
   */
  showCursor(): this {
    this.restoreCursorOnExit = false;
    this.out(ansiEscapes.cursorShow);

    return this;
  }

  /**
   * Start the console by wrapping streams and buffering output.
   */
  start(args: any[] = []): this {
    if (this.state.started) {
      return this;
    }

    this.debug('Starting console render loop');
    this.emit('start', args);
    this.wrapStreams();
    this.displayHeader();
    this.state.started = true;

    return this;
  }

  /**
   * Automatically render the console in a timeout loop at 16 FPS.
   */
  startRenderLoop() {
    if (this.isSilent() || this.isDisabled()) {
      return;
    }

    this.renderTimer = setTimeout(() => {
      this.flushOutputQueue();
    }, FPS_RATE);
  }

  /**
   * Stop the console rendering process.
   */
  stop(error: Error | null = null) {
    if (this.state.stopped) {
      return;
    }

    if (error) {
      this.debug('Stopping console with an error');
      this.errorLogs.push(...this.logs);
      this.logs = [];
    } else {
      this.debug('Stopping console render loop');
    }

    this.renderFinalOutput(error);
    this.unwrapStreams();
    this.emit('stop', [error]);
    this.state.stopped = true;
    this.state.started = false;
  }

  /**
   * Stop the background render loop.
   */
  stopRenderLoop() {
    if (this.renderTimer) {
      clearTimeout(this.renderTimer);
      this.renderTimer = null;
    }
  }

  /**
   * Unwrap the native console and reset it back to normal.
   */
  unwrapStreams() {
    if (this.isSilent() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;

      if (!this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Unwrapping `%s` stream', name);

      process[name].write = this.writers[name] as any;

      WRAPPED_STREAMS[name] = false;
    });
  }

  /**
   * Wrap the `stdout` and `stderr` streams and buffer the output as
   * to not collide with our render loop.
   */
  wrapStreams() {
    if (this.isSilent() || this.isDisabled() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;
      const stream = process[name];
      let buffer = '';

      if (this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Wrapping `%s` stream', name);

      this.bufferedStreams.push(() => {
        if (stream.isTTY && buffer) {
          this.writers[name](buffer);
        }

        buffer = '';
      });

      stream.write = (chunk: string) => {
        // No output, display immediately
        if (this.isEmptyQueue()) {
          this.writers[name](chunk);
        } else {
          buffer += String(chunk);
        }

        return true;
      };

      WRAPPED_STREAMS[name] = true;
    });
  }
}
