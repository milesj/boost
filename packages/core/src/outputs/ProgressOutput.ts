/* eslint-disable no-magic-numbers */

import chalk from 'chalk';
import cliSize from 'term-size';
import optimal, { bool, number, string } from 'optimal';
import Output from '../Output';

const STYLES = {
  bar: ['\u2588', '\u2591'],
  dash: ['=', '-'],
  hash: ['#', '-'],
  pipe: ['|', '-'],
  square: ['\u25A0', ' '],
};

export type ProgressStyle = keyof typeof STYLES;

export interface ProgressState {
  color?: boolean;
  current: number;
  style?: ProgressStyle;
  template?: string;
  total: number;
  transparent?: boolean;
}

export type ProgressRenderer = () => ProgressState;

export default class ProgressOutput extends Output<ProgressRenderer> {
  startTime: number = 0;

  stopTime: number = 0;

  protected onStart() {
    this.concurrent();
  }

  protected onFirst() {
    this.startTime = Date.now();
  }

  protected onLast() {
    this.stopTime = Date.now();
  }

  protected toString(state: ProgressState): string {
    const { color, current, style, template, total, transparent } = optimal(
      state,
      {
        color: bool(),
        current: number()
          .required()
          .gte(0),
        style: string('bar').oneOf(Object.keys(STYLES) as ProgressStyle[]),
        template: string('{percent} [{bar}] {progress}').notEmpty(),
        total: number()
          .required()
          .gt(0),
        transparent: bool(),
      },
      {
        name: 'ProgressOutput',
      },
    );

    // Mark as final for convenience
    if (current >= total) {
      this.markFinal();
    }

    // Compile our template
    const progress = Math.min(Math.max(current / total, 0.0), 1.0);
    const percent = Math.floor(progress * 100);
    const elapsed = Date.now() - this.startTime;
    const eta = percent === 100 ? 0 : elapsed * (total / current - 1);
    const rate = current / (elapsed / 1000);
    const partialTemplate = template
      .replace('{progress}', `${current}/${total}`)
      .replace('{current}', String(current))
      .replace('{elapsed}', (elapsed / 1000).toFixed(1))
      .replace('{eta}', (eta / 1000).toFixed(1))
      .replace('{percent}', `${percent.toFixed(0)}%`)
      .replace('{rate}', String(Math.round(rate)))
      .replace('{total}', String(total));

    // Render the progress bar
    const currentWidth = partialTemplate.replace('{bar}', '').length;
    const remainingWidth = cliSize().columns - currentWidth;
    const completed = Math.round(remainingWidth * progress);
    const [complete, incomplete] = STYLES[style];
    let bar = [
      complete.repeat(Math.max(0, completed)),
      (transparent ? ' ' : incomplete).repeat(Math.max(0, remainingWidth - completed)),
    ].join('');

    if (color) {
      if (percent === 100) {
        bar = chalk.green(bar);
      } else if (percent > 50) {
        bar = chalk.yellow(bar);
      } else {
        bar = chalk.red(bar);
      }
    }

    return partialTemplate.replace('{bar}', bar);
  }
}
