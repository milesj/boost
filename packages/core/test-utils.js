module.exports = require('./lib/testUtils');
