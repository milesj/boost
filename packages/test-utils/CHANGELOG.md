# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 2.2.0 - 2021-01-16

#### 🚀 Updates

- Migrate to Packemon for package building. (#135) ([1a0e9d8](https://github.com/milesj/boost/commit/1a0e9d8)), closes [#135](https://github.com/milesj/boost/issues/135)

**Note:** Version bump only for package @boost/test-utils





### 2.1.1 - 2020-11-29

#### 📦 Dependencies

- Update dev dependencies. Migrate to TypeScript v4.1. ([578d5e3](https://github.com/milesj/boost/commit/578d5e3))

**Note:** Version bump only for package @boost/test-utils





## 2.1.0 - 2020-08-17

#### 🚀 Updates

- Build packages with Rollup to support web and node targets. ([38cdad9](https://github.com/milesj/boost/commit/38cdad9))

**Note:** Version bump only for package @boost/test-utils





# 2.0.0 - 2020-07-14

#### 💥 Breaking

- Migrate to `fs.promises` API. ([944119a](https://github.com/milesj/boost/commit/944119a))
- Updated Node.js minimum requirement to v10.10. ([3719cdc](https://github.com/milesj/boost/commit/3719cdc))

#### 🚀 Updates

- Create a fixture within the temp folder. ([090e44b](https://github.com/milesj/boost/commit/090e44b))

#### 🎨 Styles

- Run Prettier. ([5cd5fc1](https://github.com/milesj/boost/commit/5cd5fc1))

#### 📘 Docs

- Update copyright years. ([1942675](https://github.com/milesj/boost/commit/1942675))
- Update license copyright year. ([e532427](https://github.com/milesj/boost/commit/e532427))

#### 📦 Dependencies

- Update final peer dependencies. ([405b8ff](https://github.com/milesj/boost/commit/405b8ff))
- **[lerna]** Update to v3.22. ([a19f27f](https://github.com/milesj/boost/commit/a19f27f))
- Update TypeScript, Jest, ESLint, and other developer packages. ([c7347a2](https://github.com/milesj/boost/commit/c7347a2))
- **[fs-extra]** Update to v9. ([0d974f8](https://github.com/milesj/boost/commit/0d974f8))

**Note:** Version bump only for package @boost/test-utils





### 1.4.1 - 2020-01-25

#### 🐞 Fixes

- Bump all packages to fix build issues. ([a8e8112](https://github.com/milesj/boost/commit/a8e8112))

**Note:** Version bump only for package @boost/test-utils





## 1.4.0 - 2019-12-07

#### 🚀 Updates

- **[PathResolver]** Add a mechanism for finding and resolving a path amongst a list of file paths or node modules. (#71) ([465e535](https://github.com/milesj/boost/commit/465e535)), closes [#71](https://github.com/milesj/boost/issues/71)

**Note:** Version bump only for package @boost/test-utils





## 1.3.0 - 2019-11-26

#### 🚀 Updates

- Support Windows OS and Node v13. (#68) ([bca6786](https://github.com/milesj/boost/commit/bca6786)), closes [#68](https://github.com/milesj/boost/issues/68)

**Note:** Version bump only for package @boost/test-utils





### 1.2.6 - 2019-11-12

#### 📘 Docs

- Fix GitHub CI badge. ([122c369](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/122c369))

#### 📦 Dependencies

- Moved `[@types](https://github.com/types)` to the root and out of packages. ([497d312](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/497d312))

#### 📋 Misc

- Add funding to all packages. ([863a614](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/863a614))

#### 🛠 Internals

- Migrate to GitHub CI and actions. (#65) ([ce59e85](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/ce59e85)), closes [#65](https://github.com/milesj/boost/tree/master/packages/test-utils/issues/65)

**Note:** Version bump only for package @boost/test-utils





### 1.2.5 - 2019-10-30

#### 📦 Dependencies

- Update minor and patch versions. ([870dd85](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/870dd85))

**Note:** Version bump only for package @boost/test-utils





### 1.2.4 - 2019-08-01

#### ⚙️ Types

- Refine types and replace `any` with `unknown`. (#58) ([43512fc](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/43512fc)), closes [#58](https://github.com/milesj/boost/tree/master/packages/test-utils/issues/58)

**Note:** Version bump only for package @boost/test-utils





### 1.2.3 - 2019-07-04

#### 📦 Dependencies

- Update fs-extra to v8.1. ([a8ddeaa](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/a8ddeaa))

#### 🛠 Internals

- Added more files to npmignore. ([16f46b8](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/16f46b8))
- Setup DangerJS and conventional changelog (#54) ([a18dd45](https://github.com/milesj/boost/tree/master/packages/test-utils/commit/a18dd45)), closes [#54](https://github.com/milesj/boost/tree/master/packages/test-utils/issues/54)

**Note:** Version bump only for package @boost/test-utils





## 1.2.0 - 2019-04-20

#### 🛠 Internals

- Decoupled `@boost/core` dependency.

### 1.0.1-1.0.4 - 2019-02-22

#### 🛠 Internals

- Updated dependencies.

# 1.0.0 - 2019-01-17

#### 🎉 Release

- Initial release!
