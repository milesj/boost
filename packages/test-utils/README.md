# Boost Testing Utils

[![Build Status](https://github.com/milesj/boost/workflows/build/badge.svg)](https://github.com/milesj/boost/actions?query=branch%3Amaster)
[![npm version](https://badge.fury.io/js/%40boost%2Ftest-utils.svg)](https://www.npmjs.com/package/@boost/test-utils)
[![npm deps](https://david-dm.org/milesj/boost.svg?path=packages/test-utils)](https://www.npmjs.com/package/@boost/test-utils)

Jest testing utilities for [Boost](https://github.com/milesj/boost/) powered applications.

## Requirements

- Boost
- Jest

## Installation

```
yarn add @boost/test-utils
```
