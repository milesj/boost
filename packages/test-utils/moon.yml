workspace:
  inheritedTasks:
    exclude: ['test', 'test-cov']
