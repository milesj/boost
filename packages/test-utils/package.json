{
  "name": "@boost/test-utils",
  "version": "4.0.0-alpha.1",
  "type": "commonjs",
  "release": "1651512178249",
  "description": "Utility classes and functions for testing Boost applications in Jest.",
  "keywords": [
    "boost",
    "test",
    "utils"
  ],
  "main": "./cjs/index.cjs",
  "types": "./dts/index.d.ts",
  "files": [
    "cjs/**/*.{cjs,mjs,map}",
    "dts/**/*.d.ts",
    "src/**/*.{ts,tsx,json}"
  ],
  "engines": {
    "node": ">=14.15.0",
    "npm": ">=6.14.0"
  },
  "repository": {
    "type": "git",
    "url": "git@github.com:milesj/boost.git",
    "directory": "packages/test-utils"
  },
  "author": "Miles Johnson",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "fs-extra": "^10.1.0"
  },
  "funding": {
    "type": "ko-fi",
    "url": "https://ko-fi.com/milesjohnson"
  },
  "packemon": {
    "api": "private",
    "format": "cjs",
    "platform": "node"
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./dts/index.d.ts",
      "node": {
        "import": "./cjs/index-wrapper.mjs",
        "require": "./cjs/index.cjs"
      }
    }
  }
}
