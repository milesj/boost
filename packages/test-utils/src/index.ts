/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './fixtures';
export * from './types';
