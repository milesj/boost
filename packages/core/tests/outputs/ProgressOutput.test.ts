import { mockConsole, mockTool } from '@boost/test-utils';
import ProgressOutput from '../../src/outputs/ProgressOutput';
import Console from '../../src/Console';

jest.mock('term-size', () => () => ({ columns: 50 }));

class TestProgressOutput extends ProgressOutput {
  testStart() {
    this.onStart();
  }

  testFirst() {
    this.onFirst();
  }

  testLast() {
    this.onLast();
  }

  testRender() {
    return this.toString(this.renderer());
  }
}

const oldDateNow = Date.now;

describe('ProgressOutput', () => {
  let output: TestProgressOutput;
  let cli: Console;

  beforeEach(() => {
    cli = mockConsole(mockTool());
    cli.render = jest.fn();

    output = new TestProgressOutput(cli, () => ({
      current: 25,
      total: 100,
    }));

    Date.now = () => 123;
  });

  afterEach(() => {
    Date.now = oldDateNow;
  });

  describe('onStart()', () => {
    it('marks as concurrent', () => {
      expect(output.isConcurrent()).toBe(false);

      output.testStart();

      expect(output.isConcurrent()).toBe(true);
    });
  });

  describe('onFirst()', () => {
    it('sets start time', () => {
      expect(output.startTime).toBe(0);

      output.testFirst();

      expect(output.startTime).not.toBe(0);
    });
  });

  describe('onLast()', () => {
    it('sets stop time', () => {
      expect(output.stopTime).toBe(0);

      output.testLast();

      expect(output.stopTime).not.toBe(0);
    });
  });

  describe('toString()', () => {
    it('renders with default params', () => {
      expect(output.testRender()).toMatchSnapshot();
    });

    it('marks as final when current hits total', () => {
      output = new TestProgressOutput(cli, () => ({
        current: 100,
        total: 100,
      }));

      output.testRender();

      expect(output.isFinal()).toBe(true);
    });

    it('renders classic style with elapsed time', () => {
      output = new TestProgressOutput(cli, () => ({
        current: 66,
        total: 100,
        style: 'classic',
        template: '{bar} | {elapsed} elapsed',
      }));

      expect(output.testRender()).toMatchSnapshot();
    });

    it('renders hash style with elapsed time', () => {
      output = new TestProgressOutput(cli, () => ({
        current: 66,
        total: 100,
        style: 'hash',
        template: '{eta} eta / {bar}',
      }));

      expect(output.testRender()).toMatchSnapshot();
    });

    it('renders pipe style with rate', () => {
      output = new TestProgressOutput(cli, () => ({
        current: 22,
        total: 100,
        style: 'pipe',
        template: '{rate} bps / {bar}',
      }));

      expect(output.testRender()).toMatchSnapshot();
    });

    it('renders square style with everything', () => {
      output = new TestProgressOutput(cli, () => ({
        current: 22,
        total: 100,
        style: 'square',
        template: '{bar} {percent} {progress} {elapsed} {eta} {rate}',
      }));

      expect(output.testRender()).toMatchSnapshot();
    });

    it('renders red when below 45 and color enabled', () => {
      output = new TestProgressOutput(cli, () => ({
        color: true,
        current: 15,
        total: 100,
        style: 'square',
        template: '{bar}',
      }));

      expect(output.testRender()).toMatchSnapshot();
    });

    it('renders yellow when above 45 and color enabled', () => {
      output = new TestProgressOutput(cli, () => ({
        color: true,
        current: 75,
        total: 100,
        style: 'square',
        template: '{bar}',
      }));

      expect(output.testRender()).toMatchSnapshot();
    });

    it('renders gren when above 90 and color enabled', () => {
      output = new TestProgressOutput(cli, () => ({
        color: true,
        current: 95,
        total: 100,
        style: 'square',
        template: '{bar}',
      }));

      expect(output.testRender()).toMatchSnapshot();
    });

    it('doesnt show incomplete progress when transparent is enabled', () => {
      output = new TestProgressOutput(cli, () => ({
        current: 50,
        total: 100,
        template: '{bar}',
        transparent: true,
      }));

      expect(output.testRender()).toMatchSnapshot();
    });
  });
});
