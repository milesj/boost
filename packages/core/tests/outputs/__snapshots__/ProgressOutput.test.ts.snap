// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressOutput toString() doesnt show incomplete progress when transparent is enabled 1`] = `"â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ                         "`;

exports[`ProgressOutput toString() renders classic style with elapsed time 1`] = `"======================------------ | 123ms elapsed"`;

exports[`ProgressOutput toString() renders gren when above 90 and color enabled 1`] = `"[32mâ– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â–   [39m"`;

exports[`ProgressOutput toString() renders hash style with elapsed time 1`] = `"64ms eta / ##########################-------------"`;

exports[`ProgressOutput toString() renders pipe style with rate 1`] = `"178.86 bps / ||||||||-----------------------------"`;

exports[`ProgressOutput toString() renders red when below 45 and color enabled 1`] = `"[31mâ– â– â– â– â– â– â– â–                                           [39m"`;

exports[`ProgressOutput toString() renders square style with everything 1`] = `"â– â– â– â–                  22% 22/100 123ms 437ms 178.86"`;

exports[`ProgressOutput toString() renders with default params 1`] = `"25% â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘â–‘ 25/100"`;

exports[`ProgressOutput toString() renders yellow when above 45 and color enabled 1`] = `"[33mâ– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â– â–             [39m"`;
